### PR TITLE
 WIP: Binary ECC: lambda projective coordinates

### PR DESCRIPTION
**Pull request by mattcaswell**
This is based on the original patch in #2386 by @bbbrumley. I've attempted to tidy it up and bring it into line with current coding style and practices in an attempt to mitigate some of the issues in #6396.

Unfortunately this is WIP because it fails in the tests. The tests passed when this patch was originally written so I did a git bisect to try and discover what commit would have caused this and I ended up at commit 920ed8c81d. Unfortunately that doesn't help a lot because that is just an improvement to the tests, to run the tests on more curves. Given that the tests were updated just a few months after this patch was written, and the patch fails even then - my guess is that the cause of the test failure stems right back to the original patch.

@bbbrumley - any ideas?

### CODE REVIEW

**Comment by bbbrumley** [comment]
Thanks, @mattcaswell ! Super impressed with your initiative.

Some of the `EC_METHOD` members are missing due to changes since 2015. That fixes some of the errors.

Best guess: Looks like I missed a custom `ec_GF2m_lambda_cmp`. So I'll implement that and get back to you.

File: crypto/ec/ec2_lambda.c [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+            || !EC_POINT_make_affine(group, r, ctx))
+        goto done;
+    if (x != NULL) {
+        if (!BN_copy(x, r->X))
+            goto done;
+        BN_set_negative(x, 0);
+    }
+    if (y != NULL) {
+        if (!BN_GF2m_add(y, r->Y, r->X)
+                || !group->meth->field_mul(group, y, y, point->X, ctx))
```
Comment (mattcaswell) [inline crypto/ec/ec2_lambda.c]: @bbbrumley Should this be `r->X` here?

**Comment by mattcaswell** [comment]
> Best guess: Looks like I missed a custom ec_GF2m_lambda_cmp. So I'll implement that and get back to you.

The same though occurred to me but it looks like `ec_GF2m_simple_cmp` converts to affine first, so shouldn't this be ok?

File: crypto/ec/ec2_lambda.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/ec/ec2_lambda.c]: Making that change seems to fix most tests except for one, for this curve:

    /* IPSec curves */
    {NID_ipsec3, &_EC_IPSEC_155_ID3.h, 0,
     "\n\tIPSec/IKE/Oakley curve #3 over a 155 bit binary field.\n"
     "\tNot suitable for ECDSA.\n\tQuestionable extension field!"},

**Comment by bbbrumley** [comment]
@mattcaswell so basically this coordinate system isn't suitable for curves with even order generators. (`NID_ipsec3` and `NID_ipsec4`). The latter was passing by chance -- I'll submit a PR for `ectest.c` that catches it.

I'll clone your personal branch and send you a PR. Also it seems pretty easy to modify `dbl` to support arbitrary curve coefficient `curve->a` so I'll do that, too.

Even after it passes the tests, I'll still want to run some security testing. Attacker gets to choose the point, so I want to see some evidence it's failing in all the right ways even when the point isn't a multiple of the generator.

Give me some time for the PR. I'm on vacation :|

**Comment by mattcaswell** [comment]
> @mattcaswell so basically this coordinate system isn't suitable for curves with even order generators. (NID_ipsec3 and NID_ipsec4). The latter was passing by chance -- I'll submit a PR for ectest.c that catches it.

Great - thanks.

>
>I'll clone your personal branch and send you a PR. Also it seems pretty easy to modify dbl to support arbitrary curve coefficient curve->a so I'll do that, too.
>
>Even after it passes the tests, I'll still want to run some security testing. Attacker gets to choose the point, so I want to see some evidence it's failing in all the right ways even when the point isn't a multiple of the generator.
>
>Give me some time for the PR. I'm on vacation :|

No problem! Enjoy your vacation!

**Comment by bbbrumley** [comment]
@mattcaswell I had some time on the train today. [This](https://github.com/bbbrumley/openssl/tree/bbb_lambda) is working, but still needs security testing. My team has some custom tooling -- I'll let you know how it goes.

**Comment by mattcaswell** [comment]
> @mattcaswell I had some time on the train today. This is working, but still needs security testing. My team has some custom tooling -- I'll let you know how it goes.

Great - thanks for working on this!

**Comment by bbbrumley** [comment]
@mattcaswell I haven't forgot this. The security testing just isn't going very well :\ The corner cases are not trivial in this coordinate system.

**Comment by mattcaswell** [comment]
> @mattcaswell I haven't forgot this. The security testing just isn't going very well :\ The corner cases are not trivial in this coordinate system.

Ok - thanks for the update

**Comment by bbbrumley** [comment]
@mattcaswell So after #6608 this coordinate system starts to show it's sharp edges. There are lots of corner cases to cover, and even then the formulas are not exception-free; point-at-infinity needs to be handled in a special way, P+Q when P=Q must take a different code path, etc.

My team is working on an alternate approach with a new `EC_METHOD`, internal to the lib so it won't affect the public API. How much of a rush should we be in to intersect 1.1.1?

**Comment by mattcaswell** [comment]
> How much of a rush should we be in to intersect 1.1.1?

Sorry for the delayed response. I've been away for a few days. The 1.1.1 release is dependent on the TLSv1.3 RFC publication - which I expect imminently (although its been imminent for quite a while now!). Once that's done there will be at least one more beta before the final 1.1.1 release. Whether there are more or not will depend on progress towards the release criteria defined here:

https://www.openssl.org/policies/releasestrat.html

I think its unlikely we're going to want to make significant changes in the EC code once we're into the final beta cycle. So I'd say time is of the essence for this change.

**Comment by romen** [comment]
#6690 provides an alternative to this PR that brings back the performance around the level of `1.1.1-pre3`, while also including the various SCA countermeasures introduced lately.

**Comment by mattcaswell** [comment]
Closing this in favour of #6690.

**Pull request closed without merging**